bentoml/yatai/deployment/azure_functions/operator.py

```diff
@@ -0,0 +1,705 @@
+        resource_group_name = f'{namespace[:29]}-{deployment_name[:60]}'
+
+    # 3-24 a-z0-9, scope: global
+    storage_account_name = f'{namespace}{deployment_name}'.lower()
+    if len(storage_account_name) > MAX_STORAGE_ACCOUNT_NAME_LENGTH:
+        storage_account_name = f'{namespace[:5]}{deployment_name[0:19]}'
+    # Replace invalid chars in storage account name to '0'
+    storage_account_name = re.sub(re.compile("[^a-z0-9]"), '0', storage_account_name)
+
+    # Azure has no documentation on the requirements for function plan name.
+    function_plan_name = deployment_name
+
+    # same as Microsoft.Web/sites
+    # 2-60, alphanumeric and hyphens. scope global
+    function_name = f'{namespace}-{deployment_name}'
+    if len(deployment_name) > MAX_FUNCTION_NAME_LENGTH:
+        function_name = f'{namespace[:19]}-{deployment_name[:40]}'
+    function_name = re.sub(re.compile("[^a-zA-Z0-9-]"), '-', function_name)
+
+    # 5-50, alphanumeric scope global
+    container_registry_name = f'{namespace}{deployment_name}'
+    if len(container_registry_name) > MAX_CONTAINER_REGISTRY_NAME_LENGTH:
+        container_registry_name = f'{namespace[:10]}{deployment_name[:40]}'
+    container_registry_name = re.sub(
+        re.compile("[^a-zA-Z0-9]"), '0', container_registry_name
+    )
+
+    return (
+        resource_group_name,
+        storage_account_name,
+        function_plan_name,
+        function_name,
+        container_registry_name,
+    )
+
+
+def _call_az_cli(command, message, parse_json=True):
+    proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    sys_default_encoding = sys.getfilesystemencoding()
+    stdout, stderr = proc.communicate()
+    if proc.returncode == 0:
+        result = stdout.decode(sys_default_encoding)
+        if result.endswith('\x1b[0m'):
+            # remove console color code: \x1b[0m
+            # https://github.com/Azure/azure-cli/issues/9903
+            result = result.replace('\x1b[0m', '')
+        logger.debug(f'AZ command "{" ".join(command)}" Result: {result}')
+        if parse_json:
+            return json.loads(result)
+        else:
+            return result
+    else:
+        error_message = stderr.decode(sys_default_encoding)
+        if not error_message:
+            error_message = stdout.decode(sys_default_encoding)
+        logger.error(f'AZ command: "{" ".join(command)}" failed: {error_message}')
+        raise AzureServiceError(f'Failed to {message}. {str(error_message)}')
+
+
+def _login_acr_registry(acr_name, resource_group_name):
+    result = _call_az_cli(
+        command=[
+            'az',
+            'acr',
+            'login',
+            '--name',
+            acr_name,
+            '--resource-group',
+            resource_group_name,
+        ],
+        message='log into Azure container registry',
+        parse_json=False,
+    ).replace('\n', '')
+    if result == 'Login Succeed':
```

Comment:
shouldn't this be `if result != 'Login Succeed':`?

The stdout matching seems off here..

Reply:
can we use docker api to actually confirm it has logged in to this registry?